types/decimal.go

```diff
@@ -0,0 +1,422 @@
+	return Dec{
+		new(big.Int).Mul(big.NewInt(i), precisionMultiplier(prec)),
+	}
+}
+
+// create a new Dec from big integer assuming whole numbers
+// CONTRACT: prec <= Precision
+func NewDecFromBigInt(i *big.Int) Dec {
+	return NewDecFromBigIntWithPrec(i, 0)
+}
+
+// create a new Dec from big integer assuming whole numbers
+// CONTRACT: prec <= Precision
+func NewDecFromBigIntWithPrec(i *big.Int, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(i, precisionMultiplier(prec)),
+	}
+}
+
+// create a new Dec from big integer assuming whole numbers
+// CONTRACT: prec <= Precision
+func NewDecFromInt(i Int) Dec {
+	return NewDecFromIntWithPrec(i, 0)
+}
+
+// create a new Dec from big integer with decimal place at prec
+// CONTRACT: prec <= Precision
+func NewDecFromIntWithPrec(i Int, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(i.BigInt(), precisionMultiplier(prec)),
+	}
+}
+
+// create a decimal from an input decimal string.
+// valid must come in the form:
+//   (-) whole integers (.) decimal integers
+// examples of acceptable input include:
+//   -123.456
+//   456.7890
+//   345
+//   -456789
+//
+// NOTE - An error will return if more decimal places
+// are provided in the string than the constant Precision.
+//
+// CONTRACT - This function does not mutate the input str.
+func NewDecFromStr(str string) (d Dec, err Error) {
+	if len(str) == 0 {
+		return d, ErrUnknownRequest("decimal string is empty")
+	}
+
+	// first extract any negative symbol
+	neg := false
+	if str[0] == '-' {
+		neg = true
+		str = str[1:]
+	}
+
+	if len(str) == 0 {
+		return d, ErrUnknownRequest("decimal string is empty")
+	}
+
+	strs := strings.Split(str, ".")
+	lenDecs := 0
+	combinedStr := strs[0]
+	if len(strs) == 2 {
+		lenDecs = len(strs[1])
+		if lenDecs == 0 || len(combinedStr) == 0 {
+			return d, ErrUnknownRequest("bad decimal length")
+		}
+		combinedStr = combinedStr + strs[1]
+	} else if len(strs) > 2 {
+		return d, ErrUnknownRequest("too many periods to be a decimal string")
+	}
+
+	if lenDecs > Precision {
+		return d, ErrUnknownRequest(
+			fmt.Sprintf("too much precision, maximum %v, len decimal %v", Precision, lenDecs))
+	}
+
+	// add some extra zero's to correct to the Precision factor
+	zerosToAdd := Precision - lenDecs
+	zeros := fmt.Sprintf(`%0`+strconv.Itoa(zerosToAdd)+`s`, "")
+	combinedStr = combinedStr + zeros
+
+	combined, ok := new(big.Int).SetString(combinedStr, 10)
+	if !ok {
+		return d, ErrUnknownRequest(fmt.Sprintf("bad string to integer conversion, combinedStr: %v", combinedStr))
+	}
+	if neg {
+		combined = new(big.Int).Neg(combined)
+	}
+	return Dec{combined}, nil
+}
+
+//______________________________________________________________________________________________
+//nolint
+func (d Dec) IsZero() bool      { return (d.Int).Sign() == 0 } // Is equal to zero
+func (d Dec) Equal(d2 Dec) bool { return (d.Int).Cmp(d2.Int) == 0 }
+func (d Dec) GT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == 1 }     // greater than
```

Comment:
I always use comparisons to zero, since the direction of comparison works on the values themselves.
(d.Int).Cmp(d2.Int) > 0 means d.Int > d2.Int, imo easier to grok visually.

Reply:
yup I like it changed